core/src/main/java/org/elasticsearch/action/support/replication/ReplicatedOperation.java

```diff
@@ -0,0 +1,334 @@
+        } else {
+            return null;
+        }
+    }
+
+    protected List<ShardRouting> getShards(ShardId shardId, ClusterState state) {
+        // can be null if the index is deleted / closed on us..
+        // nocommit: I'm not sure this is the right solution here. Deletion is easy (it doesn't matter), closing an index
+        // during replication may result in loss of acked writes. Solving this will required a bigger effort around closed
+        // indices.
+        final IndexShardRoutingTable shardRoutingTable = state.getRoutingTable().shardRoutingTableOrNull(shardId);
+        List<ShardRouting> shards = shardRoutingTable == null ? Collections.emptyList() : shardRoutingTable.shards();
+        return shards;
+    }
+
+    private void decPendingAndFinishIfNeeded() {
+        if (pendingShards.decrementAndGet() <= 0) {
+            finish();
+        }
+    }
+
+    private void finish() {
+        if (finished.compareAndSet(false, true)) {
+            final ReplicationResponse.ShardInfo.Failure[] failuresArray;
+            if (shardReplicaFailures.isEmpty()) {
+                failuresArray = ReplicationResponse.EMPTY;
+            } else {
+                failuresArray = new ReplicationResponse.ShardInfo.Failure[shardReplicaFailures.size()];
+                shardReplicaFailures.toArray(failuresArray);
+            }
+            finalResponse.setShardInfo(new ReplicationResponse.ShardInfo(
+                            totalShards.get(),
+                            successfulShards.get(),
+                            failuresArray
+                    )
+            );
+            finalResponseListener.onResponse(finalResponse);
+        }
+    }
+
+    private void finishAsFailed(Throwable throwable) {
+        if (finished.compareAndSet(false, true)) {
+            finalResponseListener.onFailure(throwable);
+        }
+    }
+
+
+    /**
+     * Should an exception be ignored when the operation is performed on the replica.
+     */
+    public static boolean ignoreReplicaException(Throwable e) {
+        if (TransportActions.isShardNotAvailableException(e)) {
+            return true;
+        }
+        // on version conflict or document missing, it means
+        // that a new change has crept into the replica, and it's fine
+        if (isConflictException(e)) {
+            return true;
+        }
+        return false;
+    }
+
+    public static boolean isConflictException(Throwable e) {
+        Throwable cause = ExceptionsHelper.unwrapCause(e);
+        // on version conflict or document missing, it means
+        // that a new change has crept into the replica, and it's fine
+        if (cause instanceof VersionConflictEngineException) {
+            return true;
+        }
+        return false;
+    }
+
+
+    interface Primary<Request extends ReplicationRequest<Request>, ReplicaRequest extends ReplicationRequest<ReplicaRequest>,
```

Comment:
I don't like Primary nor ReplicasXYZ. I prefer that we use something like ReplicationSource and ReplicationTargets. That makes it clear that the respective shards involved serve as source for the replication or as target.

I also don't think that we need two different interfaces at all. We could just provide one object to ReplicationOperation that represents the mechanisms to interact with the respective shards. With the changes I suggested before (removing term() and routingEntry()), the interface can become quite succinct with the following 4 methods:

performOnSource
failReplicationSourceShard
replicateOperation
failReplicationTargetShard